### ISSUE TITLE
Log a WARNING in case a MVC controller is annotated with @Named 

### ISSUE THREAD

**Issue by rsoika** [mvc-spec/ozark#177]
As we have already discussed [here](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/jsr371-users/-yH5KnsK88U/Vpvl3Gy0AwAJ) it does not make sense to annotate a MVC Controller class with `@Named` and a scope. 

The following conroller code works fine in Wildfly/RestEasy:

```
@Named
@RequestScoped
@Controller
@Path("/connection")
public class LoginController implements Serializable {
  String id;
  public String getId() {
	 return id;
  }
  public void setId(String id) {
	this.id = id;
  }
....
}
```

The property 'id' can be used in the view 

```
...
  <h1>#{loginController.id}</h1>
...
```

But the same code will not work for Glassfish/Jersey!

The reason is, that Jersey (or the CDI container) news the controller when rendering the view and so the properties are lost. ReastEasy (or the CDI Contianer) re-uses the bean also during the rendering phase (which makes sense as it is annotated as RequestScoped). 
But as I understand this behavior is not precisely defined by the jaxrs spec. So a implementation may re-new the controller or not. 

Of course both implementations are working fine in case a RequestScoped CDI bean is injected into the MVC Controller. I think this is exactly the same as when injecting the mvc Models Bean.

As Christian already noted in the usergroup, the JAX-RS 2.1 spec states the following in 3.1.1 "Lifecycle and Environment":

_By default a new resource class instance is created for each request to that resource. First the constructor (see Section 3.1.2) is called, then any requested dependencies are injected (see Section 3.2), then the appropriate
method (see Section 3.3) is invoked and finally the object is made available for garbage collection. An implementation MAY offer other resource class lifecycles, mechanisms for specifying these are outside the scope of this specification. E.g. an implementation based on an inversion-of-control framework may support all of the lifecycle options provided by that framework._

So for me it is now clear that the mvc controller annotated with `@Named` is not working in all Implementations (even if it works in Wildfly). My suggestion would be - if this is possible - to log a warning when an MVC controller is annotated with `@Named` or any kind of Scope.

This will help developers to not fall into this trap. In my case, the problem first occurred when I tried to deploy my application in Glassfish. And I have to refactor the complete application.

**Comment by chkal** [comment]
Just to quickly share my thoughts on this.

I would be fine with emitting warnings if controllers are annotated with `@Named`. This typically doesn't make any sense because controllers should not be directly accessed from the view.

But I'm not sure if we should really warn if a controller is annotated with a scope annotation. The JAX-RS spec explicitly allows alternative lifecycles for resource classes. And although I think that controllers should be always `@RequestScoped`, I could imagine cases where some other scope like `@SessionScoped` makes sense. So IMO a warning doesn't make sense here.

To sum up: +1 for warning in case of `@Named`, -1 for warning in case of other scope annotations.

**Comment by chkal** [comment]
Please also see #161 which is also about checking controllers and their annotations...

**Comment by rsoika** [comment]
I'm still struggling a little bit with understanding the controller concept. I repeated some tests in payara (with the ozark-jersey library 1.0.0-m3. 

You say: 

> ...controllers should not be directly accessed from the view.

I can confirm this for payara. If you try to use a controller in a view (facelet page) then this controller is renewed by some part of the framework during the render phase. So even if I annotate the controller with a scope (request or session) this does not make any sense, as the view works with a different instance as the controller instances performing the request-method. 

So for my understanding it is wrong to annotate the jax-rs controller with `@named` or `@..Scopend`.

When you develop with Wildfly you will not see this effect because for some reasons Wildfly takes care about the request- or session-scope during the complete rendering life-cycle. 

Beside this different behavior between wildfly and payara (that can lead to frustration for developers) I can not see any reason why it should make sense to annotate a MVC controller with a scope. Even if I would inject controllers into each other during a request method, I would also fear that, depending on the platform, unexpected effects may occur. 

My conclusion so far is: 

- a MVC controller should only be annotated with `@path` and `@controller` 
- the model should be injected as separate scoped CDI beans.

For that I think a warning can save developers from frustration. 

But as I said, I am not well enough familiar with the overall concept of MVC 1.0  (I have not read many books yet ;-). Maybe you have an example that shows this UseCase from a different perspective?

**Comment by chkal** [comment]
> So for my understanding it is wrong to annotate the jax-rs controller with @named or @..Scopend.

Well, I agree that the behavior of Payara is weird. It looks like Jersey uses a very "special" way for the CDI integration. IMO this behavior is not correct at all. 

However, the problems just occur because you are trying to access the controller from the view. And doing so requires you to use `@Named`. And because accessing the controller from the view is against the MVC pattern and because I don't see any other good reason for using `@Named`, I would be fine with warning the user if he uses `@Named` on a controller.

For scope annotations it is different. I could argue that `@RequestScoped` is totally fine for controllers. And the JAX-RS spec explicitly allows other lifecycles. So I don't see a good reason to warn the user in such a case. Making in controller `@SessionScoped` may seem weird, but I could imagine that users want to preserver some controller internal state between requests. So using different scope WITHOUT `@Named` is absolutely fine.

> When you develop with Wildfly you will not see this effect because for some reasons Wildfly takes care about the request- or session-scope during the complete rendering life-cycle.

In my view this is the expected and correct behavior. 

> I have not read many books yet ;-)

There are no books yet. ;-)

**Comment by chkal** [comment]
Fixed via #182
